### PR TITLE
[no ticket][risk=no] CircleCI e2e job improvement

### PR DESCRIPTION
**Pull request by aweng98**
Additional checks to avoid running tests if not neccessary:
- On master and PR branch, don't run api and ui tests if changed files are only e2e files (eg. e2e test PR merge).
- On master branch, don't run e2e job 
   * if changed files match ignore patterns
   * changed files are only e2e files (test already ran on PR so don't need to run again on merge into master branch)

### CODE REVIEW

**Review by aweng98 — COMMENTED** [review]
@calbach I'm making some more CI changes to eliminate unneccessary testing. Can you review and let me know? Not urgent.

File: .circleci/config.yml [hunk 1]
```diff
@@ -167,23 +167,44 @@ commands:
       - run:
           command: |
             if [ ${CIRCLE_BRANCH} != "" ] && [ ${CIRCLE_BRANCH} != "master" ]; then
-              diff=$(git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) -- api/src/main/resources | wc -l | xargs)
```
[the remark this answers is not in the document]
Comment (calbach) [reply .circleci/config.yml:170]: because the swagger definitions are there, which are used to generate UI code

[the remark this answers is not in the document]
Comment (aweng98) [reply .circleci/config.yml]: 👍 ah, make sense to have it. added back.

File: .circleci/choose-e2e-env.sh [hunk 2]
```diff
@@ -1,15 +1,13 @@
...
 if [[ ! -z $CIRCLE_PULL_REQUEST ]]; then
-  # Count total number of files changed
-  CHANGED_COUNT=$(git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) | wc -l | xargs)
-
-  # Count total number of e2e files
-  E2E_CHANGED_COUNT=$(git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) -- e2e | wc -l | xargs)
-
-  # If e2e files are the only files that have changed, update WORKBENCH_ENV to "test"
-  if [ $E2E_CHANGED_COUNT -gt 0 ] && [ $E2E_CHANGED_COUNT -eq $CHANGED_COUNT ]; then
+  # Count total number of changed files that are not e2e and circleci
```
Comment (calbach) [inline .circleci/choose-e2e-env.sh:6]: The new comments and logic are hard to follow, I think there are too many negatives. I've read it about 3-4 times and I'm still not entirely sure this is correct

Should the logic just be: if changes in ui/ directory (or yaml) -> run against local env, else run against test - why the focus on the test directories?

File: .circleci/config.yml [hunk 3]
```diff
@@ -183,7 +183,31 @@ commands:
...
+  halt-test-if-e2e:
+    description: "Skip non-e2e test if e2e test files are the only changed files in last commit on PR and master branch"
+    steps:
+      - run:
+          command: |
+            # Always run test on tag builds. Check is done on PR and master branch
+            if [[ -z "$CIRCLE_TAG" ]]; then
+              if [[ "$CIRCLE_BRANCH" == "master" ]]; then
+                # if no changes, exit 0. otherwise, exit 1
+                CHANGES=$(git diff HEAD~1 --quiet --name-only -- . ':!e2e')
```
Comment (calbach) [inline .circleci/config.yml:197]: I'd rather just have this run on every master merge. It is good to have this as a source of truth, especially when tests may be flaky. Optimizing on presubmit is fine though

File: .circleci/config.yml [hunk 4]
```diff
@@ -183,7 +183,31 @@ commands:
                 fi
               fi
             fi
-          name: "Halt job if code in '<< parameters.dir_names >>' directories are not changed"
+          name: "On PR branch, halt job if code in '<< parameters.dir_names >>' dir are not changed in last commit"
+
+  halt-test-if-e2e:
+    description: "Skip non-e2e test if e2e test files are the only changed files in last commit on PR and master branch"
```
Comment (calbach) [inline .circleci/config.yml:189]: Why do the other tests care about e2e? Seems like a better separation of concerns would be "API unit tests should only run when API code changes", "UI unit tests should only run when UI code changes"

File: .circleci/config.yml [hunk 5]
```diff
@@ -295,14 +320,27 @@ commands:
...
+            - run:
+                name: "Check commit message and find changed file names that do not match ignore patterns on PR branch."
+                command: bash .circleci/skip-e2e-pr-checks.sh
+      - when:
+          condition:
+            equal: [ "master", << parameters.branch >> ]
+          steps:
+            - run:
+                name: "Check for any non-e2e files that were changed in last commit on master branch."
+                command: bash .circleci/skip-e2e-master-checks.sh
```
Comment (calbach) [inline .circleci/config.yml:342]: I would prefer this ran on every master merge - or whenever we run the release, since it can fail due to environmental issues or non-obvious interactions. However, if this is being used as a step to mitigate concurrent runs, then I'm fine adding this temporarily. Running Puppeteer on a fixed interval would be another option.

File: .circleci/master-e2e-ignore-patterns.txt [hunk 6]
```diff
@@ -0,0 +1,19 @@
+.md
```
Comment (calbach) [inline .circleci/master-e2e-ignore-patterns.txt:1]: Why are these two files different?

**Review by calbach — COMMENTED** [review]
Can you provide more rationale for why you want to make these changes? In general, having e2e tests run on every merge is fine, even if there were no changes to the tests or UI code.
